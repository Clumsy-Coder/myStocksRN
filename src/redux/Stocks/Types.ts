/**
 * Data format when fetching stock quote
 * check https://iexcloud.io/docs/api/#quote
 *
 * @interface IStockQuote
 */
export interface IStockQuote {
  symbol: string; // stock ticker. ex: AAPL
  companyName: string; // company name
  primaryExchange: string; // Refers to the primary listing exchange for the symbol
  latestPrice: number; // Refers to the latest relevant price of the security which is derived from multiple sources
  latestVolume: number; // Refers to the latest total market volume of the stock across all markets
  latestUpdate: number; // latest update time in epoch represented in milliseconds
  latestTime: string; // latest update time in human readable format
  change: number; // Refers to the change in price between latestPrice and previousClose
  changePercent: number; // Refers to the percent change in price between latestPrice and previousClose
  open: number; // open stock price
  close: number; // close stock price
  high: number | null; // highest stock price during stock hours. null before 9:45 and weekends
  low: number | null; // lowest stock price during stock hours. null before 9:45 and weekends
  week52High: number; // highest stock price in the last 52 weeks
  week52Low: number; // lowest stock price in the last 52 weeks
  marketCap: number; // is calculated in real time using latestPrice
  ytdChange: number; // Refers to the price change percentage from start of year to previous close
  peRatio: number; // Refers to the price-to-earnings ratio for the company
  // volume: number; // Total volume for the stock, but only updated after market open
  avgTotalVolume: number;
  previousClose: number; // Refers to the previous trading day closing price
  extendedPrice: number; // Refers to the 15 minute delayed price outside normal market hours 0400 - 0930 ET and 1600 - 2000 ET
  extendedChange: number; // Refers to the price change between extendedPrice and latestPrice
  extendedChangePercent: number; // Refers to the price change percent between extendedPrice and latestPrice
  previousVolume: number; // Refers to the previous trading day volume

  // openTime: number;
  // closeTime: number;
  // openSource: string;
  // closeSource: string;
  // highTime: number;
  // highSource: string;
  // lowTime: number;
  // lowSource: string;

  // delayedPrice: number;
  // delayedPriceTime: number;
  // oddLotDelayedPrice: number;
  // oddLotDelayedPriceTime: number;
  // extendedPriceTime: number;
  // iexRealtimePrice: number;
  // iexRealtimeSize: number;
  // iexLastUpdated: number;
  // iexMarketPercent: number;
  // iexVolume: number;
  // iexBidPrice: number;
  // iexBidSize: number;
  // iexAskPrice: number;
  // iexAskSize: number;
  // iexOpen: null;
  // iexOpenTime: null;
  // iexClose: number;
  // iexCloseTime: number;
  // lastTradeTime: number;
  // isUSMarketOpen: Boolean;
}

/**
 * Data format when fetching stock chart data.
 * check https://iexcloud.io/docs/api/#charts
 * check https://iexcloud.io/docs/api/#historical-prices
 * check https://iexcloud.io/docs/api/#intraday-prices
 *
 * @interface IStockChart
 */
export interface IStockChart {
  date: string; // Formatted as YYYY-MM-DD
  open: number; // Adjusted data for historical dates. Split adjusted only
  close: number; // Adjusted data for historical dates. Split adjusted only
  high: number; // Adjusted data for historical dates. Split adjusted only
  low: number; // Adjusted data for historical dates. Split adjusted only
  volume: number; // Adjusted data for historical dates. Split adjusted only
  currency: string;
  change: number; // Change from previous trading day
  changePercent: number; // Change percent from previous trading day
  label: string; // A human readable format of the date depending on the range
  changeOverTime: number; // Percent change of each interval relative to first value. Useful for comparing multiple stocks
  // uClose: number; // Unadjusted data for historical dates
  // uOpen: number; // Unadjusted data for historical dates
  // uHigh: number; // Unadjusted data for historical dates
  // uLow: number; // Unadjusted data for historical dates
  // uVolume: number; // Unadjusted data for historical dates
}

// ---------------------------------------------------------------------------------------------- //
// used for Stocks reducer
export interface IStockState {
  [symbol: string]: {
    quote: {
      fetching: Boolean;
      data: IStockQuote | null;
      error?: any;
    };
    chart?: {
      fetching: boolean;
      data: IStockChart[] | null;
      error?: any;
    };
  };
}

// ---------------------------------------------------------------------------------------------- //
// action types

export enum ActionTypesEnum {
  GET_STOCK_QUOTE = 'STOCKS/GET_STOCK_QUOTE',
  GET_STOCK_QUOTE_PENDING = 'STOCKS/GET_STOCK_QUOTE_PENDING',
  GET_STOCK_QUOTE_FULFILLED = 'STOCKS/GET_STOCK_QUOTE_FULFILLED',
  GET_STOCK_QUOTE_REJECTED = 'STOCKS/GET_STOCK_QUOTE_REJECTED',

  GET_STOCK_CHART = 'STOCKS/GET_STOCK_CHART',
  GET_STOCK_CHART_PENDING = 'STOCKS/GET_STOCK_CHART_PENDING',
  GET_STOCK_CHART_FULFILLED = 'STOCKS/GET_STOCK_CHART_FULFILLED',
  GET_STOCK_CHART_REJECTED = 'STOCKS/GET_STOCK_CHART_REJECTED',
}

// ---------------------------------------------------------------------------------------------- //
// action creators

// get stock quote action creator
// called to initiate the fetching process
export interface IGetStockQuoteAC {
  type: ActionTypesEnum.GET_STOCK_QUOTE;
  readonly stockSymbol: string; // stock symbol
}

// called only in redux saga to handle side effects
export interface IGetStockQuotePendingAC {
  type: ActionTypesEnum.GET_STOCK_QUOTE_PENDING;
  readonly stockSymbol: string;
}

// called only in redux saga to handle side effects
export interface IGetStockQuoteFulfilledAC {
  type: ActionTypesEnum.GET_STOCK_QUOTE_FULFILLED;
  readonly stockSymbol: string;
  readonly payload: { data: IStockQuote };
}

// called only in redux saga to handle side effects
export interface IGetStockQuoteRejectedAC {
  type: ActionTypesEnum.GET_STOCK_QUOTE_REJECTED;
  readonly stockSymbol: string;
  readonly error: any;
}

export type StockQuoteAction =
  | IGetStockQuoteAC
  | IGetStockQuotePendingAC
  | IGetStockQuoteFulfilledAC
  | IGetStockQuoteRejectedAC;

// ---------------------------------------------------------------------------------------------- //
// get stock chart action creator
// called to initiate the fetching process
export interface IGetStockChartAC {
  type: ActionTypesEnum.GET_STOCK_CHART;
  readonly stockSymbol: string;
}

// called only in redux saga to handle side effects
export interface IGetStockChartPendingAC {
  type: ActionTypesEnum.GET_STOCK_CHART_PENDING;
  readonly stockSymbol: string;
}

// called only in redux saga to handle side effects
export interface IGetStockChartFulfilledAC {
  type: ActionTypesEnum.GET_STOCK_CHART_FULFILLED;
  readonly stockSymbol: string;
  readonly payload: { data: IStockChart[] };
}

// called only in redux saga to handle side effects
export interface IGetStockChartRejectedAC {
  type: ActionTypesEnum.GET_STOCK_CHART_REJECTED;
  readonly stockSymbol: string;
  readonly error: any;
}

export type StockChartAction =
  | IGetStockChartAC
  | IGetStockChartPendingAC
  | IGetStockChartFulfilledAC
  | IGetStockChartRejectedAC;

// ---------------------------------------------------------------------------------------------- //
// all Stock action creators
export type StockActionTypes = StockQuoteAction | StockChartAction;
