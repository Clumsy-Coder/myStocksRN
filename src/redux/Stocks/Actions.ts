/* eslint-disable max-len */
import { ActionCreator } from 'redux';

import { ActionTypes, DataDomain, Actions } from 'src/redux/Stocks/Types';

// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
//                                                                                                //
//                                  STOCK QUOTE ACTIONS CREATORS                                  //
//                                                                                                //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //

/**
 * Fetch stock quote action creator.
 * ONLY to be called by the front end.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 */
export const fetchStockQuote: ActionCreator<Actions.Quote.FetchAction> = (
  stockSymbol: string,
): Actions.Quote.FetchAction => ({
  type: ActionTypes.FETCH_STOCK_QUOTE,
  stockSymbol,
});

/**
 * Fetch stock quote pending action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 */
export const fetchStockQuotePending: ActionCreator<Actions.Quote.FetchPendingAction> = (
  stockSymbol: string,
): Actions.Quote.FetchPendingAction => ({
  type: ActionTypes.FETCH_STOCK_QUOTE_PENDING,
  stockSymbol,
});

/**
 * Fetch stock quote fulfilled action creator.
 * ONLY to be called by redux saga.
 * Only used internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param payload - data fetched
 */
export const fetchStockQuoteFulfilled: ActionCreator<Actions.Quote.FetchFulfilledAction> = (
  stockSymbol: string,
  payload: DataDomain.StockQuote,
): Actions.Quote.FetchFulfilledAction => ({
  type: ActionTypes.FETCH_STOCK_QUOTE_FULFILLED,
  stockSymbol,
  payload,
});

/**
 * Fetch stock quote rejected action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param error - error message
 */
export const fetchStockQuoteRejected: ActionCreator<Actions.Quote.FetchRejectedAction> = (
  stockSymbol: string,
  error: Error,
): Actions.Quote.FetchRejectedAction => ({
  type: ActionTypes.FETCH_STOCK_QUOTE_REJECTED,
  stockSymbol,
  error,
});

// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
//                                                                                                //
//                             STOCK DAILY ADJUSTED ACTIONS CREATORS                              //
//                                                                                                //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //

/**
 * Fetch stock Daily Adjusted action creator.
 * ONLY to be called by the front end.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param outputsize - 'compact' or 'full'
 */
export const fetchStockDailyAdj: ActionCreator<Actions.DailyAdjusted.FetchAction> = (
  stockSymbol: string,
  outputsize: 'compact' | 'full',
): Actions.DailyAdjusted.FetchAction => ({
  type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED,
  stockSymbol,
  outputsize,
});

/**
 * Fetch stock Daily Adjusted pending action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 */
export const fetchStockDailyAdjPending: ActionCreator<Actions.DailyAdjusted.FetchPendingAction> = (
  stockSymbol: string,
): Actions.DailyAdjusted.FetchPendingAction => ({
  type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED_PENDING,
  stockSymbol,
});

/**
 * Fetch Stock Daily Adjusted fulfilled action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param payload - data fetched
 */
export const fetchStockDailyAdjFulfilled: ActionCreator<Actions.DailyAdjusted.FetchFulfilledAction> = (
  stockSymbol: string,
  payload: DataDomain.StockDailyAdj,
): Actions.DailyAdjusted.FetchFulfilledAction => ({
  type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED_FULFILLED,
  stockSymbol,
  payload,
});

/**
 * Fetch Stock Daily Adjusted rejected action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param error - error message
 */
export const fetchStockDailyAdjRejected: ActionCreator<Actions.DailyAdjusted.FetchRejectedAction> = (
  stockSymbol: string,
  error: Error,
): Actions.DailyAdjusted.FetchRejectedAction => ({
  type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED_REJECTED,
  stockSymbol,
  error,
});

// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
//                                                                                                //
//                                 STOCK SEARCH ACTIONS CREATORS                                  //
//                                                                                                //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //

export const fetchStockSearch: ActionCreator<Actions.Search.FetchAction> = (
  keyword: string,
): Actions.Search.FetchAction => ({
  type: ActionTypes.SEARCH_KEYWORD,
  keyword,
});

export const fetchStockSearchPending: ActionCreator<Actions.Search.FetchPendingAction> = (
  keyword: string,
): Actions.Search.FetchPendingAction => ({
  type: ActionTypes.SEARCH_KEYWORD_PENDING,
  keyword,
});

export const fetchStockSearchFulfilled: ActionCreator<Actions.Search.FetchFulfilledAction> = (
  keyword: string,
  payload: DataDomain.StockSearch,
): Actions.Search.FetchFulfilledAction => ({
  type: ActionTypes.SEARCH_KEYWORD_FULFILLED,
  keyword,
  payload,
});

export const fetchStockSearchRejected: ActionCreator<Actions.Search.FetchRejectedAction> = (
  keyword: string,
  error: Error,
): Actions.Search.FetchRejectedAction => ({
  type: ActionTypes.SEARCH_KEYWORD_REJECTED,
  keyword,
  error,
});

// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
//                                                                                                //
//                                  STOCK BATCH ACTIONS CREATORS                                  //
//                                                                                                //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //

// /**
//  * Fetch multiple stock quote data in a single batch action creator.
//  * ONLY to be called by the front end.
//  * @param stockSymbols - An array of company stock symbol in uppercase. Ex: ['AAPL', 'SHOP]
//  */
// export const fetchStockQuoteBatch: ActionCreator<Actions.Batch.FetchQuoteBatchAction> = (): Actions.Batch.FetchQuoteBatchAction => ({
//   type: ActionTypes.FETCH_STOCK_QUOTE_BATCH,
// });

// /**
//  * Fetch multiple stock chart data in a single batch action creator.
//  * ONLY to be called by the front end.
//  * @param stockSymbols - An array of company stock symbol in uppercase. Ex: ['AAPL', 'SHOP]
//  * @param range - Date range for the stock chart
//  * @param sort - sort data by date in ascending or descending order
//  */
// export const fetchStockChartBatch: ActionCreator<Actions.Batch.FetchDailyAdjustedBatchAction> = (): Actions.Batch.FetchDailyAdjustedBatchAction => ({
//   type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED_BATCH,
// });
