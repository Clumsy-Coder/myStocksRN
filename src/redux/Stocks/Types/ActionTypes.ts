enum ActionTypes {
  FETCH_STOCK_QUOTE = 'STOCKS/FETCH_STOCK_QUOTE',
  FETCH_STOCK_QUOTE_PENDING = 'STOCKS/FETCH_STOCK_QUOTE_PENDING',
  FETCH_STOCK_QUOTE_FULFILLED = 'STOCKS/FETCH_STOCK_QUOTE_FULFILLED',
  FETCH_STOCK_QUOTE_REJECTED = 'STOCKS/FETCH_STOCK_QUOTE_REJECTED',

  FETCH_STOCK_CHART = 'STOCKS/FETCH_STOCK_CHART',
  FETCH_STOCK_CHART_PENDING = 'STOCKS/FETCH_STOCK_CHART_PENDING',
  FETCH_STOCK_CHART_FULFILLED = 'STOCKS/FETCH_STOCK_CHART_FULFILLED',
  FETCH_STOCK_CHART_REJECTED = 'STOCKS/FETCH_STOCK_CHART_REJECTED',

  SEARCH_KEYWORD = 'STOCKS/SEARCH_KEYWORD',
  SEARCH_KEYWORD_PENDING = 'STOCKS/SEARCH_KEYWORD_PENDING',
  SEARCH_KEYWORD_FULFILLED = 'STOCKS/SEARCH_KEYWORD_FULFILLED',
  SEARCH_KEYWORD_REJECTED = 'STOCKS/SEARCH_KEYWORD_REJECTED',

  SET_SEARCH_KEYWORD = 'STOCKS/SET_SEARCH_KEYWORD',
  CLEAR_SEARCH_KEYWORD = 'STOCKS/CLEAR_SEARCH_KEYWORD',

  SET_STOCK_METADATA = 'STOCKS/SET_STOCK_METADATA',

  FETCH_STOCK_QUOTE_BATCH = 'STOCKS/FETCH_STOCK_QUOTE_BATCH',
  // FETCH_STOCK_DAILY_ADJUSTED_BATCH = 'STOCKS/FETCH_STOCK_DAILY_ADJUSTED_BATCH',
}

export default ActionTypes;
