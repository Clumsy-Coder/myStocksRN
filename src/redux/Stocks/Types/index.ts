import ActionTypes from '@redux/Stocks/Types/ActionTypes';
import * as Actions from '@redux/Stocks/Types/Actions';
import * as DataDomain from '@redux/Stocks/Types/DataDomain';
import * as Selectors from '@redux/Stocks/Types/Selectors';
import * as Reducer from '@redux/Stocks/Types/Reducer';

export { ActionTypes };
export { Actions };
export { DataDomain };
export { Selectors };
export { Reducer };
