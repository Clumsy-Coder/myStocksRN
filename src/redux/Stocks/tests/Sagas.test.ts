import { runSaga } from 'redux-saga';
import sinon from 'sinon';
import { AxiosResponse } from 'axios';
import { takeEvery } from 'redux-saga/effects';
import { expectSaga } from 'redux-saga-test-plan';
import * as matchers from 'redux-saga-test-plan/matchers';

import sagaWatcher, {
  fetchStockQuoteSaga,
  fetchStockDailyAdjustedSaga,
} from 'src/redux/Stocks/Sagas';
import { ActionTypes, StocksActions, DataDomain, Actions } from 'src/redux/Stocks/Types';
import * as actions from 'src/redux/Stocks/Actions';
import * as api from 'src/share/Utilities';

const stockSymbol1 = 'IBM';
const stockSymbol2 = 'AAPL';
const stockSymbol3 = 'SHOP';
const stockQuoteData1: DataDomain.StockQuote = {
  'Global Quote': {
    '01. symbol': 'IBM',
    '02. open': '119.3700',
    '03. high': '119.4650',
    '04. low': '117.5900',
    '05. price': '118.3900',
    '06. volume': '4179906',
    '07. latest trading day': '2020-05-22',
    '08. previous close': '119.1200',
    '09. change': '-0.7300',
    '10. change percent': '-0.6128%',
  },
};

const stockQuoteData2: DataDomain.StockQuote = {
  'Global Quote': {
    '01. symbol': 'AAPL',
    '02. open': '315.7700',
    '03. high': '319.2300',
    '04. low': '315.3500',
    '05. price': '318.8900',
    '06. volume': '20240356',
    '07. latest trading day': '2020-05-22',
    '08. previous close': '316.8500',
    '09. change': '2.0400',
    '10. change percent': '0.6438%',
  },
};

const stockQuoteData3: DataDomain.StockQuote = {
  'Global Quote': {
    '01. symbol': 'SHOP',
    '02. open': '804.6300',
    '03. high': '826.3799',
    '04. low': '793.6400',
    '05. price': '825.1700',
    '06. volume': '2761628',
    '07. latest trading day': '2020-05-22',
    '08. previous close': '802.3500',
    '09. change': '22.8200',
    '10. change percent': '2.8441%',
  },
};

const stockDailyAdjData1: DataDomain.StockDailyAdj = {
  'Meta Data': {
    '1. Information': 'Daily Time Series with Splits and Dividend Events',
    '2. Symbol': 'IBM',
    '3. Last Refreshed': '2020-05-22',
    '4. Output Size': 'Compact',
    '5. Time Zone': 'US/Eastern',
  },
  'Time Series (Daily)': {
    '2020-05-22': {
      '1. open': '119.3700',
      '2. high': '119.4650',
      '3. low': '117.5900',
      '4. close': '118.3900',
      '5. adjusted close': '118.3900',
      '6. volume': '4179906',
      '7. dividend amount': '0.0000',
      '8. split coefficient': '1.0000',
    },
    '2020-05-21': {
      '1. open': '120.9900',
      '2. high': '121.7200',
      '3. low': '118.9700',
      '4. close': '119.1200',
      '5. adjusted close': '119.1200',
      '6. volume': '4018329',
      '7. dividend amount': '0.0000',
      '8. split coefficient': '1.0000',
    },
  },
};

const stockDailyAdjData2: DataDomain.StockDailyAdj = {
  'Meta Data': {
    '1. Information': 'Daily Time Series with Splits and Dividend Events',
    '2. Symbol': 'AAPL',
    '3. Last Refreshed': '2020-05-22',
    '4. Output Size': 'Compact',
    '5. Time Zone': 'US/Eastern',
  },
  'Time Series (Daily)': {
    '2020-05-22': {
      '1. open': '315.7700',
      '2. high': '319.2300',
      '3. low': '315.3500',
      '4. close': '318.8900',
      '5. adjusted close': '318.8900',
      '6. volume': '20240356',
      '7. dividend amount': '0.0000',
      '8. split coefficient': '1.0000',
    },
    '2020-05-21': {
      '1. open': '318.6600',
      '2. high': '320.8900',
      '3. low': '315.8700',
      '4. close': '316.8500',
      '5. adjusted close': '316.8500',
      '6. volume': '25672211',
      '7. dividend amount': '0.0000',
      '8. split coefficient': '1.0000',
    },
  },
};

const stockSearchData: DataDomain.StockSearch = {
  bestMatches: [
    {
      '1. symbol': 'IBM',
      '2. name': 'International Business Machines Corporation',
      '3. type': 'Equity',
      '4. region': 'United States',
      '5. marketOpen': '09:30',
      '6. marketClose': '16:00',
      '7. timezone': 'UTC-05',
      '8. currency': 'USD',
      '9. matchScore': '1.0000',
    },
    {
      '1. symbol': 'IBMM',
      '2. name': 'iShares iBonds Dec 2024 Term Muni Bond ETF',
      '3. type': 'ETF',
      '4. region': 'United States',
      '5. marketOpen': '09:30',
      '6. marketClose': '16:00',
      '7. timezone': 'UTC-05',
      '8. currency': 'USD',
      '9. matchScore': '0.8571',
    },
  ],
};

const stockSearchData2: DataDomain.StockSearch = {
  bestMatches: [
    {
      '1. symbol': 'AAPL',
      '2. name': 'Apple Inc.',
      '3. type': 'Equity',
      '4. region': 'United States',
      '5. marketOpen': '09:30',
      '6. marketClose': '16:00',
      '7. timezone': 'UTC-05',
      '8. currency': 'USD',
      '9. matchScore': '1.0000',
    },
    {
      '1. symbol': 'AAPL.ARG',
      '2. name': 'Apple Inc.',
      '3. type': 'Equity',
      '4. region': 'Argentina',
      '5. marketOpen': '11:00',
      '6. marketClose': '17:00',
      '7. timezone': 'UTC-03',
      '8. currency': 'ARS',
      '9. matchScore': '0.7273',
    },
  ],
};

describe('Stocks Saga', () => {
  describe('Stocks quote watcher', () => {
    it(`Should watch ${ActionTypes.FETCH_STOCK_QUOTE} action`, () => {
      const gen = sagaWatcher();
      const expected = takeEvery(ActionTypes.FETCH_STOCK_QUOTE, fetchStockQuoteSaga);
      const actual = gen.next().value;

      expect(actual).toEqual(expected);
    });

    it(`Should fire on ${ActionTypes.FETCH_STOCK_QUOTE} action in case success`, () => {
      return expectSaga(sagaWatcher)
        .provide([
          [matchers.call.fn(api.fetchStockQuoteUrl), Promise.resolve({ data: stockQuoteData1 })],
        ])
        .put(actions.fetchStockQuotePending(stockSymbol1))
        .call(api.fetchStockQuoteUrl, stockSymbol1)
        .put(actions.fetchStockQuoteFulfilled(stockSymbol1, stockQuoteData1))
        .not.put(actions.fetchStockQuoteRejected(stockSymbol1, new Error('')))
        .dispatch(actions.fetchStockQuote(stockSymbol1))
        .silentRun()
        .then((result) => expect(result.toJSON()).toMatchSnapshot());
    });

    it(`Should fire on ${ActionTypes.FETCH_STOCK_QUOTE} action in case failure`, () => {
      return expectSaga(sagaWatcher)
        .provide([[matchers.call.fn(api.fetchStockQuoteUrl), Promise.reject(new Error(''))]])
        .put(actions.fetchStockQuotePending(stockSymbol1))
        .call(api.fetchStockQuoteUrl, stockSymbol1)
        .put(actions.fetchStockQuoteRejected(stockSymbol1, new Error('')))
        .not.put(actions.fetchStockQuoteFulfilled(stockSymbol1, stockQuoteData1))
        .dispatch(actions.fetchStockQuote(stockSymbol1))
        .silentRun()
        .then((result) => expect(result.toJSON()).toMatchSnapshot());
    });
  });

  describe('Stocks quote saga', () => {
    it('Should load and handle Stock quote data in case of success', async () => {
      const dispatchedActions: StocksActions[] = [];
      const fakeStore = {
        getState: (): {} => ({}),
        dispatch: (action: StocksActions): number => dispatchedActions.push(action),
      };
      const promiseResponse: AxiosResponse<DataDomain.StockQuote> = {
        data: stockQuoteData1,
        headers: {},
        status: 200,
        statusText: '',
        request: {},
        config: {},
      };

      const stub = sinon.stub(api, 'fetchStockQuoteUrl').returns(Promise.resolve(promiseResponse));
      await runSaga(fakeStore, fetchStockQuoteSaga, actions.fetchStockQuote(stockSymbol1));

      expect(stub.calledOnce).toBe(true);
      expect(dispatchedActions.length).toEqual(2);
      expect(dispatchedActions[0]).toEqual(actions.fetchStockQuotePending(stockSymbol1));
      expect(dispatchedActions[1]).toEqual(
        actions.fetchStockQuoteFulfilled(stockSymbol1, stockQuoteData1),
      );

      stub.restore(); // important: do NOT remove
    });

    it('Should load and handle Stock quote data in case of failure', async () => {
      const dispatchedActions: StocksActions[] = [];
      const fakeStore = {
        getState: (): {} => ({}),
        dispatch: (action: StocksActions): number => dispatchedActions.push(action),
      };

      const stub = sinon.stub(api, 'fetchStockQuoteUrl').returns(Promise.reject(new Error('')));
      await runSaga(fakeStore, fetchStockQuoteSaga, actions.fetchStockQuote(stockSymbol1));

      expect(stub.calledOnce).toBe(true);
      expect(dispatchedActions.length).toEqual(2);
      expect(dispatchedActions[0]).toEqual(actions.fetchStockQuotePending(stockSymbol1));
      expect(dispatchedActions[1]).toEqual(
        actions.fetchStockQuoteRejected(stockSymbol1, new Error('')),
      );

      stub.restore(); // important: do NOT remove
    });
  });

  describe('Stocks chart saga', () => {
    it('Should load and handle Stock chart data in case of success', async () => {
      const dispatchedActions: StocksActions[] = [];
      const fakeStore = {
        getState: (): {} => ({}),
        dispatch: (action: StocksActions): number => dispatchedActions.push(action),
      };
      const promiseResponse: AxiosResponse<DataDomain.StockDailyAdj> = {
        data: stockDailyAdjData1,
        headers: {},
        status: 200,
        statusText: '',
        request: {},
        config: {},
      };

      const stub = sinon
        .stub(api, 'fetchStockDailyAdjustedUrl')
        .returns(Promise.resolve(promiseResponse));
      await runSaga(
        fakeStore,
        fetchStockDailyAdjustedSaga,
        actions.fetchStockDailyAdj(stockSymbol1, 'compact'),
      );

      expect(stub.calledOnce).toBe(true);
      expect(dispatchedActions.length).toEqual(2);
      expect(dispatchedActions[0]).toEqual(actions.fetchStockDailyAdjPending(stockSymbol1));
      expect(dispatchedActions[1]).toEqual(
        actions.fetchStockDailyAdjFulfilled(stockSymbol1, stockDailyAdjData1),
      );

      stub.restore(); // important: do NOT remove
    });

    it('Should load and handle Stock chart data in case of failure', async () => {
      const dispatchedActions: StocksActions[] = [];
      const fakeStore = {
        getState: (): {} => ({}),
        dispatch: (action: StocksActions): number => dispatchedActions.push(action),
      };

      const stub = sinon
        .stub(api, 'fetchStockDailyAdjustedUrl')
        .returns(Promise.reject(new Error('')));
      await runSaga(
        fakeStore,
        fetchStockDailyAdjustedSaga,
        actions.fetchStockDailyAdj(stockSymbol1, 'compact'),
      );

      expect(stub.calledOnce).toBe(true);
      expect(dispatchedActions.length).toEqual(2);
      expect(dispatchedActions[0]).toEqual(actions.fetchStockDailyAdjPending(stockSymbol1));
      expect(dispatchedActions[1]).toEqual(
        actions.fetchStockDailyAdjRejected(stockSymbol1, new Error('')),
      );

      stub.restore(); // important: do NOT remove
    });
  });
});
