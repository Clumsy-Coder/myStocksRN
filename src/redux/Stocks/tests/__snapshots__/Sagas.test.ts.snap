// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stocks Saga Stocks quote watcher Should fire on STOCKS/FETCH_STOCK_QUOTE action in case failure 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "STOCKS/FETCH_STOCK_QUOTE",
          "@@redux-saga-test-plan/json/function/fetchStockQuoteSaga",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "STOCKS/FETCH_STOCK_DAILY_ADJUSTED",
          "@@redux-saga-test-plan/json/function/fetchStockDailyAdjustedSaga",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "stockSymbol": "IBM",
            "type": "STOCKS/FETCH_STOCK_QUOTE",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchStockQuoteSaga",
      },
      "type": "FORK",
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "STOCKS/FETCH_STOCK_QUOTE",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "STOCKS/FETCH_STOCK_DAILY_ADJUSTED",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "STOCKS/FETCH_STOCK_QUOTE",
      },
      "type": "TAKE",
    },
  ],
}
`;

exports[`Stocks Saga Stocks quote watcher Should fire on STOCKS/FETCH_STOCK_QUOTE action in case success 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "STOCKS/FETCH_STOCK_QUOTE",
          "@@redux-saga-test-plan/json/function/fetchStockQuoteSaga",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "STOCKS/FETCH_STOCK_DAILY_ADJUSTED",
          "@@redux-saga-test-plan/json/function/fetchStockDailyAdjustedSaga",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "stockSymbol": "IBM",
            "type": "STOCKS/FETCH_STOCK_QUOTE",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchStockQuoteSaga",
      },
      "type": "FORK",
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "STOCKS/FETCH_STOCK_QUOTE",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "STOCKS/FETCH_STOCK_DAILY_ADJUSTED",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "STOCKS/FETCH_STOCK_QUOTE",
      },
      "type": "TAKE",
    },
  ],
}
`;
